api/envoy/service/ext_proc/v3/external_processor.proto

```diff
@@ -362,14 +362,47 @@ message HeaderMutation {
   repeated string remove_headers = 2;
 }
 
-// Replace the entire message body chunk received in the corresponding
-// HttpBody message with this new body, or clear the body.
+// The body response message corresponding to MXN body mode.
+message MxnBodyResponse {
+  // Envoy sends this body mutation out as it is.
+  bytes body = 1;
+
+  // Set this flag to true for the last body response.
+  bool end_of_body = 2;
+
+  // The ext_proc server sends this number to notify Envoy how many body chunks are received.
+  // The server may send a response message with ``chunks_received`` set to 1 for each body chunk it receives.
+  // It may also choose to wait for receiving multiple body chunks, then send an aggregated number back to Envoy for confirmation.
+  //
+  // After receives this, Envoy will
+  //
+  // * Stop the waiting timer in its state machine to avoid timeout events.
+  // * Consume that amount of chunks in it's queue to avoid flow control being triggered. This also enables more data chunks
+  //   be able to get into Envoy.
+  //
+  // In the situation when the ext_proc server needs to buffer large amount of data before sending back body responses, to avoid
+  // Envoy timeouts or flow control events being triggered, it should send standalone ``MxnBodyResponse`` messages with
+  // ``chunks_received`` field set back to Envoy.
+  //
+  // In the case the ext_proc server does not need to buffer large amount of data, and buffering time is short,  the server
+  // may choose to send a ``MxnBodyResponse`` message with both ``body`` and ``chunks_received`` field set.
+  // This can save some gRPC calls thus is a more efficient and recommended way for sending confirmation in such situation.
+  uint32 chunks_received = 3;
```

Comment:
I think this behavior is problematic. There is a maximum number of chunks that Envoy can buffer before it will stop processing data. But ext_proc developer does not know this maximum. As such the only behavior that guarantees correct behavior is acknowledge every chunk. Which makes this field pointless.

What is the value of this behavior that the ext_proc server has to clear Envoy's internal buffers? Why is it needed?

Reply:
For synchronous mode, the ext_proc <---> side stream server flow control is done by this check:

https://github.com/envoyproxy/envoy/blob/b4e356885ef56134c3c8531d1c0d0f05a74151ad/source/extensions/filters/http/ext_proc/processor_state.cc#L392

This flow control is basically try to avoid a fast clients jam a slow server.  We are not planning to remove such flow control in the 1st phase.

Reply:
We can relax the limit to be a larger, and fixed number, like 32MB, for MXN body mode.  Then the server will know when to start send back such confirmations with the data it receives..

Reply:
After some offline discussion with @tyxia about the new flow control added recently,  I decided to remove the queue based flow control for MXN mode, thus no need the bytes tracking and this "chunks_received" server response.  This simplify the API and protocol a lot.